### PR TITLE
Add load balancer and retry stats bar charts to clients

### LINKED ISSUE #936: Add additional client stats to admin dashboard
Including load balancer pool status (available/size) and retry budget.

### PR DESCRIPTION
**Pull request by rmars**
Fixes #936

* Adds Retries stats to the top router stats bar
* Adds bar chart of retry budget remaining
* Adds bar chart of load balancer availability

Retries calcluated as:
retries/total (across all paths) + retries/requeues (across all clients) / requests (total across all paths)

screen shot 2017-01-27 at 1 15 01 pm

Follow ups:
* [x] we should get the retry budget from the config (default: 20%) 
* The rgb values from the css are what I got from a color dripper of Sterling's mocks and should probably be replaced with hex values

### CODE REVIEW

**Comment by siggy** [comment]
this looks awesome! a few notes on the ui:

- consider moving the numbers away from the start and end of the bar. it's confusing to me to see a bar/gauge with a `1` on both ends. maybe just one text element that says `1/1 load balancers available`
- i'm not grocking the 3-color gauge. is it just meant to show a gradient? if so, i'd say just make it a gradient.
- i'm seeing a `-` in place of a number on the retries chart. does this mean the metric is absent? if so, does it make sense to default to 1?
- when i make the window wide, it looks a little bit weird that a big gap opens up between the bar charts. not sure the best approach here, perhaps pin retries to the left? or grow retries with the success rate graph? i'm not sure if either of those make sense, just thinking out loud.

File: admin/src/main/resources/io/buoyant/admin/js/src/router_client.js [hunk 1]
```diff
@@ -99,10 +103,37 @@ define([
       return summary;
     }
 
-    return function (metricsCollector, routers, client, $metricsEl, routerName, $chartEl, colors, $toggleLinks, shouldExpandInitially) {
+    function barChartColorFn(percent) {
+      return percent < 0.5 ? "orange" : "green";
+    }
+
+    function barPercentCalc(data) {
```
Comment (esbie) [inline admin/src/main/resources/io/buoyant/admin/js/src/router_client.js]: I'd like to see these barchart-related methods in their own function or object named `lbBarChart` to help separate them from the rest of the router client. Currently it's hard to tell whether or not these methods rely on the router client's scope. I'd start by defining them outside of `RouterClient`.

Same situation for `RouterSummary` and `retriesBarChart`

File: admin/src/main/resources/io/buoyant/admin/js/src/router_client.js [hunk 2]
```diff
@@ -99,10 +103,37 @@ define([
...
+
+    return function (metricsCollector, routers, client, $container, routerName, colors, shouldExpandInitially) {
       var metricPartial = Handlebars.compile(metricPartialTemplate);
       Handlebars.registerPartial('metricPartial', metricPartial);
 
+      var $contentContainer = $container.find(".client-content-container");
+
+      var $metricsEl = $container.find(".metrics-container");
+      var $chartEl = $container.find(".chart-container");
+      var $toggleLinks = $container.find(".client-toggle");
```
Comment (esbie) [inline admin/src/main/resources/io/buoyant/admin/js/src/router_client.js]: these 3 vars don't look used?

Comment (rmars) [inline admin/src/main/resources/io/buoyant/admin/js/src/router_client.js]: I refactored these - they were being passed in originally:
```
return function (metricsCollector, routers, client, $metricsEl, routerName, $chartEl, colors, $toggleLinks, shouldExpandInitially)
```

They're used later on in the function - maybe github has cut the rest of the function off:

```
    return function (metricsCollector, routers, client, $container, routerName, colors, shouldExpandInitially) {
      var metricPartial = Handlebars.compile(metricPartialTemplate);
      Handlebars.registerPartial('metricPartial', metricPartial);

      var $contentContainer = $container.find(".client-content-container");

      var $metricsEl = $container.find(".metrics-container");
      var $chartEl = $container.find(".chart-container");
      var $toggleLinks = $container.find(".client-toggle");
      var $lbBarChart = $container.find(".lb-bar-chart");

      var clientColor = colors.color;
      var latencyLegend = createLatencyLegend(colors.colorFamily);
      var metricDefinitions = getMetricDefinitions(routerName, client.label);

      var $expandLink = $toggleLinks.find(".client-expand");
      var $collapseLink = $toggleLinks.find(".client-collapse");

      renderMetrics($metricsEl, client, [], [], clientColor);
      var chart = SuccessRateGraph($chartEl.find(".client-success-rate"), colors.color);
```

Comment (esbie) [inline admin/src/main/resources/io/buoyant/admin/js/src/router_client.js]: ah 👍

File: admin/src/main/resources/io/buoyant/admin/js/src/router_summary.js [hunk 3]
```diff
@@ -115,7 +118,8 @@ define([
       var pathQuery = Query.pathQuery().allPaths().withRouter(routerName).withMetrics(["requests", "retries/total"]).build();
 
       var $retriesBarChart = $barChartEl.find(".retries-bar-chart");
-      var retriesBarChart = new BarChart($retriesBarChart, getBarChartColor);
+      var barChartHelpers = getRetryBarChartHelpers();
+      var retriesBarChart = new BarChart($retriesBarChart, barChartHelpers.getColor);
```
Comment (esbie) [inline admin/src/main/resources/io/buoyant/admin/js/src/router_summary.js]: OK this is great. I think we can make it even greater by tweaking how the chart is instantiated. Let's make RetriesBarChart more of a wrapper that contains the scope and less of a chart library. Here's the interface I'm thinking:
```
var retriesBarChart = new RetriesBarChart($retriesBarChart)
```
```
function RetryBarChart($container) {
  function getColor(...){...}
  function getPercent(...) {...}
  ...
  var chart = new BarChart($container, getColor);
  return {update: chart.update};
}
```

File: admin/src/main/resources/io/buoyant/admin/js/spec/BarChartSpec.js [hunk 4]
```diff
@@ -6,49 +6,112 @@ define([
...
+      expect($barContainer.attr('class')).toContain("red");
       expect($barContainer.width()).toBe(containerWidth);
-      expect($bar.width()).toBe(Math.round(60/200 * containerWidth));
+      expect($bar.width() / $barContainer.width()).toBeCloseTo(60/200);
+    });
 
-      lbBarChart.update(generateData(90, 100));
-      $bar = $lbContainer.find(".overlay-bars.bar");
-      $barContainer = $lbContainer.find(".overlay-bars.bar-container");
+    it("udpates the bar's color when new data is padded in", function() {
```
Comment (esbie) [inline admin/src/main/resources/io/buoyant/admin/js/spec/BarChartSpec.js]: tests are very clean now, good work!

File: admin/src/main/resources/io/buoyant/admin/js/src/router_summary.js (hunk 3 above)
Comment (rmars) [inline admin/src/main/resources/io/buoyant/admin/js/src/router_summary.js]: omg I like this a lot!

File: admin/src/main/resources/io/buoyant/admin/js/src/router_client.js [hunk 5]
```diff
@@ -7,14 +7,46 @@ define([
...
+
+      var numer = data["loadbalancer/available"] || {};
+      var denom = data["loadbalancer/size"] || {};
+      var percent = (!denom || !denom.value) ? 0 : (numer.value || 0) / denom.value;
+
+      return {
+        percent: percent,
+        label: {
+          description: "Endpoints available",
+          value: (numer.value || "-") + " / " + (denom.value || "-")
```
Comment (adleong) [inline admin/src/main/resources/io/buoyant/admin/js/src/router_client.js]: it seems like 0s display as "-".  I think a 0 should display as "0".

Comment (rmars) [inline admin/src/main/resources/io/buoyant/admin/js/src/router_client.js]: oops yeah, agreed!

**Pull request merged**